external/storm-rocketmq/src/main/java/org/apache/storm/rocketmq/spout/RocketMQSpout.java

```diff
@@ -0,0 +1,189 @@
+        Validate.notEmpty(properties, "Consumer properties can not be empty");
+        boolean ordered = getBoolean(properties, RocketMQConfig.CONSUMER_MESSAGES_ORDERLY, false);
+
+        int queueSize = getInteger(properties, SpoutConfig.QUEUE_SIZE, ObjectReader.getInt(conf.get(Config.TOPOLOGY_MAX_SPOUT_PENDING)));
+        queue = new LinkedBlockingQueue<>(queueSize);
+
+        // Since RocketMQ Consumer is thread-safe, RocketMQSpout uses a single
+        // consumer instance across threads to improve the performance.
+        synchronized (RocketMQSpout.class) {
+            if (consumer == null) {
+                consumer = new DefaultMQPushConsumer();
+                RocketMQConfig.buildConsumerConfigs(properties, (DefaultMQPushConsumer)consumer);
+
+                if (ordered) {
+                    consumer.registerMessageListener(new MessageListenerOrderly() {
+                        @Override
+                        public ConsumeOrderlyStatus consumeMessage(List<MessageExt> msgs,
+                                                                   ConsumeOrderlyContext context) {
+                            if (process(msgs)) {
+                                return ConsumeOrderlyStatus.SUCCESS;
+                            } else {
+                                return ConsumeOrderlyStatus.SUSPEND_CURRENT_QUEUE_A_MOMENT;
+                            }
+                        }
+                    });
+                } else {
+                    consumer.registerMessageListener(new MessageListenerConcurrently() {
+                        @Override
+                        public ConsumeConcurrentlyStatus consumeMessage(List<MessageExt> msgs,
+                                                                        ConsumeConcurrentlyContext context) {
+                            if (process(msgs)) {
+                                return ConsumeConcurrentlyStatus.CONSUME_SUCCESS;
+                            } else {
+                                return ConsumeConcurrentlyStatus.RECONSUME_LATER;
+                            }
+                        }
+                    });
+                }
+
+                try {
+                    consumer.start();
+                } catch (MQClientException e) {
+                    throw new RuntimeException(e);
+                }
+            }
+        }
+
+        int maxRetry = getInteger(properties, SpoutConfig.MESSAGES_MAX_RETRY, SpoutConfig.DEFAULT_MESSAGES_MAX_RETRY);
+        int ttl = getInteger(properties, SpoutConfig.MESSAGES_TTL, SpoutConfig.DEFAULT_MESSAGES_TTL);
+        this.messageRetryManager = new DefaultMessageRetryManager(queue, maxRetry, ttl);
+        this.collector = collector;
+    }
+
+    public boolean process(List<MessageExt> msgs) {
+        if (msgs.isEmpty()) {
+            return true;
+        }
+        MessageSet messageSet = new MessageSet(msgs);
```

Comment:
any plans of offering a TupleMapper to flatten the schema or you want to preserve the same data in rocketmq and send it downstream